pkg/core/writer/correlations/client.go

```diff
@@ -80,92 +108,163 @@ func NewCorrelationClient(ctx context.Context, conf *config.WriterConfig) (Corre
 	}, nil
 }
 
 func (cc *Client) putRequestOnChan(r *request) error {
+	// prevent requests against empty dimension names and values
+	if r.DimName == "" || r.DimValue == "" {
+		// logging this as debug because this means there's no actual dimension to correlate with
+		// and because this isn't being taken off on the request sender and subject to retries, this could
+		// potentially spam the logs
+		atomic.AddInt64(&cc.TotalInvalidDimensions, int64(1))
+		log.WithFields(log.Fields{"method": r.operation, "correlation": r.Correlation}).Debug("No dimension key or value to correlate to")
+		return nil
+	}
+
+	r.contextWithCancel = &contextWithCancel{}
+	r.ctx, r.cancel = context.WithCancel(requestcounter.ContextWithRequestCounter(context.Background()))
+
 	var err error
 	select {
 	case cc.requestChan <- r:
 	case <-cc.ctx.Done():
 		err = context.DeadlineExceeded
 	default:
-		err = errChFull
+		err = ErrChFull
 	}
 	return err
 }
 
-func (cc *Client) correlateCb(r *request, _ []byte, _ error) {
-	if cc.logUpdates {
-		log.WithFields(log.Fields{
-			"method":      http.MethodPut,
-			"correlation": r.Correlation,
-		}).Info("Updated dimension")
+func (cc *Client) putRequestOnRetryChan(r *request) error {
+	// handle request counter
+	if requestcounter.GetRequestCount(r.ctx) == cc.maxAttempts {
+		return errMaxAttempts
 	}
-}
+	requestcounter.IncrementRequestCount(r.ctx)
 
-func (cc *Client) Correlate(cor *Correlation) {
-	err := cc.putRequestOnChan(&request{Correlation: cor, operation: http.MethodPut, callback: cc.correlateCb})
-	if err != nil && err != context.DeadlineExceeded {
-		log.WithError(err).WithFields(log.Fields{
-			"method":      http.MethodPut,
-			"correlation": cor,
-		}).Error("Unable to update dimension, not retrying")
+	// set the time to retry
+	r.sendAt = cc.now().Add(cc.sendDelay)
+
+	if r.ctx.Err() != nil {
+		return errRequestCancelled
 	}
+
+	var err error
+	select {
+	case <-r.ctx.Done():
+		err = errRequestCancelled
+	case cc.retryChan <- r:
+	case <-cc.ctx.Done():
+		err = context.DeadlineExceeded
+	default:
+		err = errRetryChFull
+	}
+
+	return err
 }
 
-func (cc *Client) deleteCb(r *request, _ []byte, _ error) {
-	if cc.logUpdates {
-		log.WithFields(log.Fields{
-			"method":      http.MethodDelete,
-			"correlation": r.Correlation,
-		}).Info("Updated dimension")
+// CorrelateCB is a call back invoked with Correlate requests
+// it is not invoked if the reqeust is deduplicated, cancelled, or the client context is cancelled
+type CorrelateCB func(cor *Correlation, err error)
+
+// Correlate
+func (cc *Client) Correlate(cor *Correlation, cb CorrelateCB) {
+	err := cc.putRequestOnChan(&request{
+		Correlation: cor,
+		operation:   http.MethodPut,
+		callback: func(body []byte, statuscode int, err error) {
+			switch statuscode {
+			case http.StatusOK:
+				if cc.logUpdates {
+					log.WithFields(log.Fields{"method": http.MethodPut, "correlation": cor}).Info("Updated dimension")
+				}
+			case http.StatusTeapot:
```

Comment:
Does our backend really return this or is this just some internal hack you are using?

Reply:
It will in the not too distant future.  The scope of my bug creeped to include this now.